perfkitbenchmarker/linux_benchmarks/cloud_spanner_ycsb_benchmark.py

```diff
@@ -0,0 +1,204 @@
+flags.DEFINE_string('cloud_spanner_instance',
+                    'ycsb-' + getpass.getuser(),
+                    'The Cloud Spanner instance used in the YCSB benchmark.')
+flags.DEFINE_string('cloud_spanner_database',
+                    'ycsb',
+                    'The Cloud Spanner database used in the YCSB benchmark.')
+flags.DEFINE_enum('cloud_spanner_readmode',
+                  'query', ['query', 'read'],
+                  'The Cloud Spanner read mode used in the YCSB benchmark.')
+flags.DEFINE_integer('cloud_spanner_boundedstaleness',
+                     0,
+                     'The Cloud Spanner bounded staleness used in the YCSB '
+                     'benchmark.')
+flags.DEFINE_integer('cloud_spanner_batchinserts',
+                     1,
+                     'The Cloud Spanner batch inserts used in the YCSB '
+                     'benchmark.')
+flags.DEFINE_string('cloud_spanner_custom_ycsb_release',
+                    None,
+                    'If provided, the URL of a custom YCSB release')
+flags.DEFINE_string('cloud_spanner_desp',
+                    'YCSB',
+                    'The description of the Cloud Spanner instance.')
+flags.DEFINE_integer('cloud_spanner_nodes',
+                     1,
+                     'The number of nodes for the Cloud Spanner instance.')
+flags.DEFINE_string('cloud_spanner_config',
+                    'regional-us-central1',
+                    'The config for the Cloud Spanner instance.')
+flags.DEFINE_string('cloud_spanner_ddl',
+                    """CREATE TABLE usertable (
+                    id STRING(MAX), field0 STRING(MAX), field1 STRING(MAX),
+                    field2 STRING(MAX), field3 STRING(MAX), field4 STRING(MAX),
+                    field5 STRING(MAX), field6 STRING(MAX), field7 STRING(MAX),
+                    field8 STRING(MAX), field9 STRING(MAX),
+                    ) PRIMARY KEY(id)""",
+                    'The schema DDL for the Cloud Spanner database.')
+
+
+def GetConfig(user_config):
+  config = configs.LoadConfig(BENCHMARK_CONFIG, user_config, BENCHMARK_NAME)
+  if FLAGS['ycsb_client_vms'].present:
+    config['vm_groups']['default']['vm_count'] = FLAGS.ycsb_client_vms
+  return config
+
+
+def CheckPrerequisites(benchmark_config):
+  for scope in REQUIRED_SCOPES:
+    if scope not in FLAGS.gcloud_scopes:
+      raise ValueError('Scope {0} required.'.format(scope))
+
+
+def Prepare(benchmark_spec):
+  """Prepare the virtual machines to run cloud spanner benchmarks.
+
+  Args:
+    benchmark_spec: The benchmark specification. Contains all data that is
+        required to run the benchmark.
+  """
+  benchmark_spec.always_call_cleanup = True
+
+  benchmark_spec.spanner_instance = gcp_spanner.GcpSpannerInstance(
+      FLAGS.cloud_spanner_instance, FLAGS.cloud_spanner_desp,
+      FLAGS.cloud_spanner_nodes, FLAGS.cloud_spanner_config,
+      FLAGS.cloud_spanner_database, FLAGS.cloud_spanner_ddl)
```

Comment:
IMHO, DRY is irrelevant here. gcp_spanner.py is a class library. It's sufficient for a class library to expose configurable parameters at class constructor level. At the class library level, we don't necessarily know what a user will do. A user of the class library may choose to use FLAGS, which happens to be the case of cloud_spanner_ycsb_benchmarks.py, or it may also choose to use a configuration file, or simply define them as constants should they be fixed for a specific benchmark. As said, these parameters may change from benchmarks to benchmarks. And it makes sense that each benchmark owns the default values for those parameters (again, benchmarks can own the default values as FLAG, configuration file, constants, etc). Otherwise, when a user runs a benchmark, he/she is now forced to always pass in FLAGS, which is not convenient.